net: lwm2m: add gateway_obj prefix check to lwm2m_message_handler

walzsi opened the pull request:
The idea of the lwm2m gateway object 25 is that messages can be forwarded from a "gateway" to a connected "childdevice". The server can reach any device managed by the gateway by prefixing the LWM2M CoAP request with the prefix announced by the object. To be able to handle these messages, a callback is implemented in the message handler which forwards these messages if required.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -2307,6 +2310,47 @@ static int handle_request(struct coap_packet *request, struct lwm2m_message *msg
 	}
 #endif
 
+#if defined(CONFIG_LWM2M_GATEWAY_OBJ_SUPPORT)
```
I'd rather see this code encapsulated into some helper function (likely implemented along the gateway object implementation) and only called here. I'm not in favor of bloating generic code with object-specific stuff.
For example something like `lwm2m_gw_request_handle()` which could return 0 if the message was consumed (was prefixed), 1 if the engine should process it as a regular message (unprefixed) or negative on error.

rlubos reviewed (commented):
For the compliance issues, make sure your `user.name` and `user.email` git configs match the ones you used in the signnoff.

rlubos commented:
FYI @rerickson1 since you seem to have been involved in GW object introduction.

SeppoTakalo reviewed (changes requested):
It is unclear to me, how these Gateway objects are supposed to work.

Spec: https://www.openmobilealliance.org/release/LwM2M_Gateway/V1_1-20210518-A/HTML-Version/OMA-TS-LWM2M_Gateway-V1_1-20210518-A.html

Basically, it should allow us to publish LwM2M objects that are not related to this device under some specific prefix.
However, it worried me that who and how is now in responsible of handling the content formatting and CoAP transmission.
The spec is quite clear that all those READ, WRITE, EXECUTE should work on paths that are under the prefix.

So if the callback in Gateway Object is supposed to do all that, then my proposal is as follows:
* Create new API in `lwm2m_obj_gateway.c` called `lwm2m_gw_handle_req(struct coap_packet *request, struct lwm2m_message *msg)`
* Move all this path matching into the `lwm2m_gw_handle_req()` so it does not clutter the `lwm2m_message_handling.c`
* Call this inside `handle_request()`
```c
if (IS_ENABLED(CONFIG_LWM2M_GATEWAY_OBJ_SUPPORT)) {
   if (lwm2m_gw_handle_req(request, msg) == 0) {
        return 0;
    }
```
* `lwm2m_gw_handle_req()` should only return 0 if it completely handles the request. It should take care all. Set up all the content format, writers, coap response packets etc.

rerickson1 commented:
> It is unclear to me, how these Gateway objects are supposed to work.

The spec leaves it up to the implementation to do all the work.

We successfully implemented this on BLE end nodes.
The gateway needed to intercept LwM2M messages and see if there was a prefix. If there was, we used callbacks to then strip off the prefix and forward the CoAP message over BLE.
On the BLE end node we ran the LwM2M client and changed the LwM2M engine so it wasnt dependent on sockets. We made an abstraction layer in the engine so you could hook in any transport for sending/receiving the LwM2M CoAP messages.

I can point anyone to all the open source code that made this happen. Getting proper gateway support in the Zephyr tree would be great.
The other piece is having a server that supports it. Has Leshan made any headway yet?

Your proposal sounds like it could work.

walzsi committed:
net: lwm2m: add gateway callback to handle prefixed messages

Adding a callback for handling lwm2m messages with prefixed paths defined
by the gateway object. If CONFIG_LWM2M_GATEWAY_OBJ_SUPPORT is set,
each path is checked for the prefix stored in the object instances of the
gateway object 25. If prefixes match the msg is passed to the gw_msg_cb.

Signed-off-by: Simon Walz <simon.walz@autosen.com>

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -2290,6 +2293,12 @@ static int handle_request(struct coap_packet *request, struct lwm2m_message *msg
 		goto error;
 	}
 
+	if (IS_ENABLED(CONFIG_LWM2M_GATEWAY_OBJ_SUPPORT)) {
+		if (lwm2m_gw_handle_req(msg) == 0) {
+				return 0;
+			}
+	}
```
```suggestion
	if (IS_ENABLED(CONFIG_LWM2M_GATEWAY_OBJ_SUPPORT)) {
		if (lwm2m_gw_handle_req(msg) == 0) {
				return 0;
		}
	}
```

Whitespace error.

SeppoTakalo replied:
Another thing: If you think that callbacks should not automatically respond with messages, then I think we should call here `lwm2m_reset_message(msg, true);` and then return something negative so that the `lwm2m_udp_receive()` does not assume the `msg` is now a response.

For example, if we release message, we could call `return -ENOMSG` here.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -64,6 +64,9 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #ifdef CONFIG_LWM2M_RW_SENML_CBOR_SUPPORT
 #include "lwm2m_rw_senml_cbor.h"
 #endif
+#ifdef CONFIG_LWM2M_GATEWAY_OBJ_SUPPORT
+#include "lwm2m_obj_gateway.h"
+#endif
```
```suggestion
#include "lwm2m_obj_gateway.h"
```

This can be unconditional. the `IS_ENABLED()`  doesn't call the API if GW object is not enabled, but it needs the header for compiler to check the call.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_obj_gateway.h:
```diff
@@ -17,4 +19,9 @@
 #define LWM2M_GATEWAY_IOT_DEVICE_OBJECTS_RID 3
 /* clang-format on */
 
+typedef int (*lwm2m_engine_gateway_msg_cb)(struct lwm2m_message *msg);
+
+void lwm2m_register_gw_callback(uint16_t obj_inst_id, lwm2m_engine_gateway_msg_cb cb);
```
```suggestion
typedef int (*lwm2m_engine_gateway_msg_cb)(struct lwm2m_message *msg);

void lwm2m_register_gw_callback(uint16_t obj_inst_id, lwm2m_engine_gateway_msg_cb cb);
```

This can now be considered as part of the public API and deserves some documentation.
It is still unclear whose responsibility is to call `lwm2m_reset_message(msg, true);` that will release the message structure.

Also, is the callback supposed to fill up the `msg` object with the response message? At least Ack is required. Please see `lwm2m_udp_receive()`.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_obj_gateway.c:
```diff
@@ -146,6 +147,57 @@ static struct lwm2m_engine_obj_inst *lwm2m_gw_create(uint16_t obj_inst_id)
+			return ret;
+		}
+
+		ret = gateway_msg_cb[obj_inst_id](msg);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+	return 0;
```
After a second thought, do we need to return here one of the 3 options that might have happened:
1. GW object did not handle the request. This is now signalled as negative code.
2. Callback handled the request, but did not return response.
3. Callback handled the requested and returned a response in `msg`.

Now only option 1 is handled, but I'm unsure if we assume 2 or 3 option when we return 0.
Please check `

walzsi replied:
i've thought about it again and i think it makes sense to leave the transmission of the msg to the lwm2m engine. 
The engine is the "gateway" and therefore, in my opinion, it should also take over the transmission. The cb handles the complete coap msg and its formatting. This also allows us to reuse the existing coap responses for certain error codes.

SeppoTakalo reviewed (changes requested):
Thanks you for the fixes looks better now.

I think there is now one question that should be solved. I left the comments inline.

In short:
* It should be clear who calls `lwm2m_reset_message()` on these GW queries.
* Responses from these queries should be handled somehow.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_obj_gateway.c:
```diff
@@ -146,6 +147,58 @@ static struct lwm2m_engine_obj_inst *lwm2m_gw_create(uint16_t obj_inst_id)
+		return ret;
+	}
+
+	for (int obj_inst_id = 0; obj_inst_id == CONFIG_LWM2M_GATEWAY_MAX_INSTANCES - 1;
+	     obj_inst_id++) {
+		void *prefix = NULL;
+		uint16_t prefix_len = 0;
+
+		obj_inst = get_engine_obj_inst(LWM2M_OBJECT_GATEWAY_ID, obj_inst_id);
+		if (obj_inst == NULL) {
+			return -ENOENT;
+		}
+		lwm2m_get_res_buf(
+			&LWM2M_OBJ(LWM2M_OBJECT_GATEWAY_ID, obj_inst_id, LWM2M_GATEWAY_PREFIX_RID),
+			&prefix, NULL, &prefix_len, NULL);
+
+		if (strncmp(options[0].value, (char *)prefix, prefix_len) != 0 || prefix == NULL) {
+			return -ENOENT;
+		}
+
+		if (gateway_msg_cb[obj_inst_id] == NULL) {
+			return -ENOENT;
+		}
+
+		/* Delete prefix from path*/
+		ret = coap_options_to_path(&options[1], ret - 1, &msg->path);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = gateway_msg_cb[obj_inst_id](msg);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+	return 0;
```
```suggestion
	for (int index = 0; index < MAX_INSTANCE_COUNT; index++) {
		/* Skip uninitialized objects */
		if (!inst[index].obj) {
			continue;
		}

		char *prefix = device_table[index].prefix;
		size_t prefix_len = strlen(prefix);

		if (prefix_len != options[0].len) {
			continue;
		}
		if (strncmp(options[0].value, prefix, prefix_len) != 0) {
			continue;
		}
		
		if (gateway_msg_cb[index] == NULL) {
			return -ENOENT;
		}

		/* Delete prefix from path*/
		ret = coap_options_to_path(&options[1], ret - 1, &msg->path);
		if (ret < 0) {
			return ret;
		}

		return gateway_msg_cb[index](msg);
	}
	return -ENOENT;
```

I recommend that you don't deal with object_ids in the loop, as those are not necessary same as index in the array.
Just use indexes directly. Then if callback is responsible of dealing the `msg` object, call it when found. If we exit the loop, we know that none of the gateway objects matched.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_obj_gateway.c:
```diff
@@ -146,6 +147,58 @@ static struct lwm2m_engine_obj_inst *lwm2m_gw_create(uint16_t obj_inst_id)
+
+void lwm2m_register_gw_callback(uint16_t obj_inst_id, lwm2m_engine_gateway_msg_cb cb)
+{
+	gateway_msg_cb[obj_inst_id] = cb;
```
You need to check here what is the index of given `obj_inst_id`.
They don't always match.

Loop for the `lwm2m_gw_create()` for how to find the instance ID.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_obj_gateway.h:
```diff
@@ -17,4 +19,23 @@
 #define LWM2M_GATEWAY_IOT_DEVICE_OBJECTS_RID 3
 /* clang-format on */
 
+typedef int (*lwm2m_engine_gateway_msg_cb)(struct lwm2m_message *msg);
+
+/**
+ * @brief Register a callback which handles the prefixed messages from the server.
+ *  
+ * The callback gets triggered each time the prefix in front of a recieved lwm2m 
+ * msg path matches the prefix set in the LWM2M_GATEWAY_PREFIX_RID buffer.
+ * 
+ * It must handle the content of the coap message completely. 
+ * In case of success the LwM2M engine will then send the formatted coap message, 
+ * otherwise a coap response code is sent.
+ * @param[in] obj_inst_id Object instance ID generating the callback
+ * @param[in] cb gateway message callback
+ * @return  0 for success, -ENOENT if the message should be handled by the lwm2m
+ *          engine or negative error codes in case of error.
+ */
+void lwm2m_register_gw_callback(uint16_t obj_inst_id, lwm2m_engine_gateway_msg_cb cb);
+int lwm2m_gw_handle_req(struct lwm2m_message *msg);
+
```
```suggestion
/**
 * @brief A callback which handles the prefixed messages from the server.
 *  
 * The callback gets triggered each time the prefix in front of a recieved lwm2m 
 * msg path matches the prefix set in the LWM2M_GATEWAY_PREFIX_RID buffer.
 * 
 * It must handle the content of the coap message completely. 
 * In case of success the LwM2M engine will then send the formatted coap message, 
 * otherwise a coap response code is sent.
 *
 * Example of returning CoAP response:
 * @code{.c}
 * lwm2m_init_message(msg);
 * // Write CoAP packet to msg->out.out_cpkt
 * return 0;
 * @endcode
 *
 *
 * @return 0 if msg contains a  valid CoAP response.
 * @return  negative error code otherwise. 
 */
typedef int (*lwm2m_engine_gateway_msg_cb)(struct lwm2m_message *msg);

/**
 * @brief Register a callback which handles the prefixed messages from the server.
 *  
 */
void lwm2m_register_gw_callback(uint16_t obj_inst_id, lwm2m_engine_gateway_msg_cb cb);

/**
 * @brief Check if given message is handled by Gateway callback.
 * 
 * @return 0 if msg was handled by Gateawy and contains a valid response. Negative error code otherwise.
 * @return -ENOENT if this msg was not handled by Gateway object.
 */
int lwm2m_gw_handle_req(struct lwm2m_message *msg);

```

SeppoTakalo reviewed (commented):
This looks OK now.

I noticed that two cases are not handled properly, so I left my suggstions inline:
* If Gateway object instance ID is not same as its index in the array.
* If prefix length is not the same length as first part of CoAP path. This might match empty prefix as well.

SeppoTakalo commented:
From my point of view, this looks OK now.
But I would like @rerickson1 to give a final thought on this, as he has been using the GW object before.
Hopefully this does not conflict the previous usecase.

I think this now makes a sense. Gateway callback may create the CoAP response as they wish and the LwM2M engine actually handles sending&retrying it back.
I hope someone can test this and verify if the code that I suggested is actually working. I would even like if someone can show some examples of its use.

rerickson1 commented:
> From my point of view, this looks OK now. But I would like @rerickson1 to give a final thought on this, as he has been using the GW object before. Hopefully this does not conflict the previous usecase.
> 
> I think this now makes a sense. Gateway callback may create the CoAP response as they wish and the LwM2M engine actually handles sending&retrying it back. I hope someone can test this and verify if the code that I suggested is actually working. I would even like if someone can show some examples of its use.

Agree, this is good for now.

To make this useful, there needs to be be a change to abstract the LwM2M engine away from sockets.
Formalizing a PR based on something like this: https://github.com/LairdCP/zephyr/commit/46c0c9f784c9c94d447df6189672ec23dea17fc2

rlubos reviewed (commented):
@walzsi Please have a final look at the compliance, there are a few nits reported.
